Geometry/GEMGeometryBuilder/plugins/GEMGeometryValidate.cc

```diff
@@ -0,0 +1,316 @@
+  vector<float> pitch_;
+  vector<float> stripslen_;
+  string infileName_;
+  string outfileName_;
+  int tolerance_;
+};
+
+GEMGeometryValidate::GEMGeometryValidate(const edm::ParameterSet& iConfig)
+    : infileName_(
+          iConfig.getUntrackedParameter<string>("infileName", "cmsRecoGeom-2021.root")),  //it was cmsGeom10.root
+      outfileName_(iConfig.getUntrackedParameter<string>("outfileName", "validateGEMGeometry.root")),
+      tolerance_(iConfig.getUntrackedParameter<int>("tolerance", 6)) {
+  fwGeometry_.loadMap(infileName_.c_str());
+  outFile_ = new TFile(outfileName_.c_str(), "RECREATE");
+}
+
+void GEMGeometryValidate::analyze(const edm::Event& event, const edm::EventSetup& eventSetup) {
+  eventSetup.get<MuonGeometryRecord>().get(gemGeometry_);
+
+  if (gemGeometry_.isValid()) {
+    LogVerbatim("GEMGeometry") << "Validating GEM chamber geometry";
+
+    validateGEMChamberGeometry();
+
+    validateGEMEtaPartitionGeometry();
+
+  } else
+    LogVerbatim("GEMGeometry") << "Invalid GEM geometry";
+}
+
+void GEMGeometryValidate::validateGEMChamberGeometry() {
+  clearData();
+
+  for (auto const& it : gemGeometry_->chambers()) {
+    GEMDetId chId = it->id();
+    GlobalPoint gp = it->surface().toGlobal(LocalPoint(0.0, 0.0, 0.0));
+
+    const TGeoMatrix* matrix = fwGeometry_.getMatrix(chId.rawId());
+
+    if (!matrix) {
+      LogVerbatim("GEMGeometry") << "Failed to get matrix of GEM chamber with detid: " << chId.rawId();
+      continue;
+    }
+    compareTransform(gp, matrix);
+
+    auto const& shape = fwGeometry_.getShapePars(chId.rawId());
+
+    if (!shape) {
+      LogVerbatim("GEMGeometry") << "Failed to get shape of GEM chamber with detid: " << chId.rawId();
+      continue;
+    }
+    compareShape(it, shape);
+  }
+  makeHistograms("GEM Chamber");
+}
+
+void GEMGeometryValidate::validateGEMEtaPartitionGeometry() {
+  clearData2();
+
+  for (auto const& it : gemGeometry_->etaPartitions()) {
+    GEMDetId chId = it->id();
+    const int n_strips = it->nstrips();
+    const float n_pitch = it->pitch();
+    const StripTopology& topo = it->specificTopology();
+    const float stripLen = topo.stripLength();
+    const float* parameters = fwGeometry_.getParameters(chId.rawId());
+    nstrips_.push_back(fabs(n_strips - parameters[0]));
+
+    if (n_strips) {
+      for (int istrips = 1; istrips <= n_strips; istrips++) {
+        pitch_.push_back(fabs(n_pitch - parameters[2]));
+
+        stripslen_.push_back(fabs(stripLen - parameters[1]));
+      }
+    } else {
+      LogVerbatim("GEMGeometry") << "ATTENTION! nStrips == 0";
+    }
+  }
+  makeHistograms2("GEM Eta Partition");
+}
+
+void GEMGeometryValidate::compareTransform(const GlobalPoint& gp, const TGeoMatrix* matrix) {
+  double local[3] = {0.0, 0.0, 0.0};
+  double global[3];
+
+  matrix->LocalToMaster(local, global);
+
+  float distance = getDistance(GlobalPoint(global[0], global[1], global[2]), gp);
+  if ((distance >= 0.0) && (distance < 1.0e-7))
```

Comment:
```suggestion
  if (distance < 1.0e-7)
```

Reply:
Do you want to allow negative distances? Then you need:
`if (std::abs(distance) < 1.0e-7)`